sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java

```diff
@@ -0,0 +1,291 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
+import java.security.*;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.*;
+
+public class TLSEvaluationStage implements OfferEvaluationStage {
+
+    private static final Logger LOGGER = (Logger) LoggerFactory.getLogger(TLSEvaluationStage.class);
+
+    private String serviceName;
+    private String taskName;
+    private CertificateAuthorityClient certificateAuthorityClient;
+    private SecretsClient secretsClient;
+    private KeyPairGenerator keyPairGenerator;
+
+    public TLSEvaluationStage(
+            String serviceName, String taskName, CertificateAuthorityClient certificateAuthorityClient, SecretsClient secretsClient, KeyPairGenerator keyPairGenerator) {
+        this.serviceName = serviceName;
+        this.taskName = taskName;
+        this.certificateAuthorityClient = certificateAuthorityClient;
+        this.secretsClient = secretsClient;
+        this.keyPairGenerator = keyPairGenerator;
+    }
+
+    public static TLSEvaluationStage fromEnvironmentForService(String serviceName, String taskName) throws IOException, InvalidKeySpecException {
+
+        SchedulerFlags flags = SchedulerFlags.fromEnv();
+
+        KeyFactory keyFactory = null;
+        try {
+            keyFactory = KeyFactory.getInstance("RSA");
+        } catch (NoSuchAlgorithmException e) {
+            e.printStackTrace();
+        }
+
+        PemReader pemReader = new PemReader(new StringReader(flags.getServiceAccountPrivateKeyPEM()));
+        PrivateKey privateKey = keyFactory.generatePrivate(
+                new PKCS8EncodedKeySpec(pemReader.readPemObject().getContent()));
+
+        ServiceAccountIAMTokenProvider serviceAccountIAMTokenProvider = new ServiceAccountIAMTokenProvider.Builder()
+                .setIamUrl(URLHelper.fromUnchecked(DcosConstants.IAM_AUTH_URL))
+                .setUid(flags.getServiceAccountUid())
+                .setPrivateKey((RSAPrivateKey) privateKey)
+                .build();
+        TokenProvider tokenProvider = new CachedTokenProvider(serviceAccountIAMTokenProvider);
+
+        Executor executor = Executor.newInstance(
+                new HttpClientBuilder()
+                        .setTokenProvider(tokenProvider)
+                        .setLogger(LOGGER)
+                        .setRedirectStrategy(new LaxRedirectStrategy())
+                        .build());
+
+        CertificateAuthorityClient certificateAuthorityClient = new DefaultCAClient(executor);
+        SecretsClient secretsClient = new DefaultSecretsClient(executor);
+
+        KeyPairGenerator keyPairGenerator = null;
+        try {
+            keyPairGenerator = KeyPairGenerator.getInstance("RSA");
+        } catch (NoSuchAlgorithmException e) {
+            e.printStackTrace();
```

Comment:
I don't know what is the best way to handle this as I don't think there is a chance that this will throw an exception.